File: backend/src/main/java/com/project/backend/service/product/ProductService.java
package com.project.backend.service.product;

import com.project.backend.dto.request.product.ImageUpsertRequest;
import com.project.backend.dto.request.product.ProductCreateRequest;
import com.project.backend.dto.request.product.ProductUpdateRequest;
import com.project.backend.dto.request.product.VariantUpsertRequest;
import com.project.backend.dto.response.product.ImageResponse;
import com.project.backend.dto.response.product.ProductDetailResponse;
import com.project.backend.dto.response.product.ProductSummaryResponse;
import com.project.backend.dto.response.product.VariantResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface ProductService {

    // ── Product CRUD ─────────────────────────────────────────────────────────

    /** Tạo sản phẩm mới (kèm ít nhất 1 variant trong req). */
    ProductDetailResponse createProduct(ProductCreateRequest req);

    /** Cập nhật thông tin cơ bản của sản phẩm (name, slug, description, category, isActive). */
    ProductDetailResponse updateProduct(Integer id, ProductUpdateRequest req);

    /** Bật/tắt trạng thái hiển thị của sản phẩm (isActive = true | false). */
    void setActive(Integer id, Boolean isActive);

    /** Lấy chi tiết sản phẩm theo id (dùng cho admin). */
    ProductDetailResponse getById(Integer id);

    /** Lấy chi tiết sản phẩm theo slug (dùng cho khách hàng). */
    ProductDetailResponse getBySlug(String slug);

    /** Danh sách sản phẩm đang active, phân trang (dùng cho storefront). */
    Page<ProductSummaryResponse> listActive(Pageable pageable);

    /** Danh sách toàn bộ sản phẩm, phân trang (dùng cho admin). */
    Page<ProductSummaryResponse> listAdmin(Pageable pageable);

    // ── Variant operations ───────────────────────────────────────────────────

    /** Thêm biến thể mới vào sản phẩm. */
    VariantResponse addVariant(Integer productId, VariantUpsertRequest req);

    /** Cập nhật thông tin biến thể (price, basePrice, size). */
    VariantResponse updateVariant(Integer variantId, VariantUpsertRequest req);

    /** Xóa biến thể theo id. */
    void deleteVariant(Integer variantId);

    // ── Image operations ─────────────────────────────────────────────────────

    /** Thêm ảnh vào biến thể. */
    ImageResponse addImage(Integer variantId, ImageUpsertRequest req);

    /** Đặt ảnh này làm ảnh chính của variant (is_main = true), reset ảnh khác về false. */
    ImageResponse setMainImage(Integer imageId);

    /** Xóa ảnh theo id. */
    void deleteImage(Integer imageId);
}

File: frontend/src/services/productService.js
import axiosInstance from './axiosInstance';

/**
 * Product Service - Xử lý tất cả API calls liên quan đến sản phẩm
 */

const productService = {
  /**
   * Lấy danh sách sản phẩm đang active (cho storefront)
   * @param {Object} params - { page, size, sort }
   * @returns {Promise} Page<ProductSummaryResponse>
   */
  getProducts: async (params = {}) => {
    const { page = 0, size = 12, sort = 'id,desc' } = params;
    const response = await axiosInstance.get('/api/products', {
      params: { page, size, sort }
    });
    return response.data;
  },

  /**
   * Lấy chi tiết sản phẩm theo slug
   * @param {string} slug - Product slug
   * @returns {Promise} ProductDetailResponse
   */
  getProductBySlug: async (slug) => {
    const response = await axiosInstance.get(`/api/products/${slug}`);
    return response.data;
  },

  /**
   * Tìm kiếm sản phẩm theo category
   * @param {number} categoryId - Category ID
   * @param {Object} params - { page, size, sort }
   * @returns {Promise} Page<ProductSummaryResponse>
   */
  getProductsByCategory: async (categoryId, params = {}) => {
    const { page = 0, size = 12, sort = 'id,desc' } = params;
    const response = await axiosInstance.get('/api/products', {
      params: { categoryId, page, size, sort }
    });
    return response.data;
  },

  // ========== ADMIN APIs ==========

  /**
   * Lấy danh sách tất cả sản phẩm (bao gồm inactive) - Admin only
   * @param {Object} params - { page, size, sort }
   * @returns {Promise} Page<ProductSummaryResponse>
   */
  getAllProducts: async (params = {}) => {
    const { page = 0, size = 20, sort = 'id,desc' } = params;
    const response = await axiosInstance.get('/api/admin/products', {
      params: { page, size, sort }
    });
    return response.data;
  },

  /**
   * Lấy chi tiết sản phẩm theo ID - Admin only
   * @param {number} id - Product ID
   * @returns {Promise} ProductDetailResponse
   */
  getProductById: async (id) => {
    const response = await axiosInstance.get(`/api/admin/products/${id}`);
    return response.data;
  },

  /**
   * Tạo sản phẩm mới - Admin only
   * @param {Object} productData - ProductCreateRequest
   * @returns {Promise} ProductDetailResponse
   */
  createProduct: async (productData) => {
    const response = await axiosInstance.post('/api/admin/products', productData);
    return response.data;
  },

  /**
   * Cập nhật sản phẩm - Admin only
   * @param {number} id - Product ID
   * @param {Object} productData - ProductUpdateRequest
   * @returns {Promise} ProductDetailResponse
   */
  updateProduct: async (id, productData) => {
    const response = await axiosInstance.put(`/api/admin/products/${id}`, productData);
    return response.data;
  },

  /**
   * Bật/tắt trạng thái sản phẩm - Admin only
   * @param {number} id - Product ID
   * @param {boolean} isActive - true/false
   * @returns {Promise}
   */
  setProductActive: async (id, isActive) => {
    const response = await axiosInstance.patch(
      `/api/admin/products/${id}/active`,
      null,
      { params: { value: isActive } }
    );
    return response.data;
  },

  // ========== VARIANT APIs ==========

  /**
   * Thêm variant mới vào sản phẩm - Admin only
   * @param {number} productId - Product ID
   * @param {Object} variantData - VariantUpsertRequest
   * @returns {Promise} VariantResponse
   */
  addVariant: async (productId, variantData) => {
    const response = await axiosInstance.post(
      `/api/admin/products/${productId}/variants`,
      variantData
    );
    return response.data;
  },

  /**
   * Cập nhật variant - Admin only
   * @param {number} variantId - Variant ID
   * @param {Object} variantData - VariantUpsertRequest
   * @returns {Promise} VariantResponse
   */
  updateVariant: async (variantId, variantData) => {
    const response = await axiosInstance.put(
      `/api/admin/variants/${variantId}`,
      variantData
    );
    return response.data;
  },

  /**
   * Xóa variant - Admin only
   * @param {number} variantId - Variant ID
   * @returns {Promise}
   */
  deleteVariant: async (variantId) => {
    const response = await axiosInstance.delete(`/api/admin/variants/${variantId}`);
    return response.data;
  },

  // ========== IMAGE APIs ==========

  /**
   * Thêm ảnh vào variant - Admin only
   * @param {number} variantId - Variant ID
   * @param {Object} imageData - ImageUpsertRequest { image, isMain }
   * @returns {Promise} ImageResponse
   */
  addImage: async (variantId, imageData) => {
    const response = await axiosInstance.post(
      `/api/admin/variants/${variantId}/images`,
      imageData
    );
    return response.data;
  },

  /**
   * Đặt ảnh làm ảnh chính - Admin only
   * @param {number} imageId - Image ID
   * @returns {Promise} ImageResponse
   */
  setMainImage: async (imageId) => {
    const response = await axiosInstance.patch(`/api/admin/images/${imageId}/main`);
    return response.data;
  },

  /**
   * Xóa ảnh - Admin only
   * @param {number} imageId - Image ID
   * @returns {Promise}
   */
  deleteImage: async (imageId) => {
    const response = await axiosInstance.delete(`/api/admin/images/${imageId}`);
    return response.data;
  }
};

export default productService;
